File: backend/agents/skillGapAgent.js
const { callGroq } = require('../services/groqService');
const { parseJSON, normalizeSkillGap } = require('../utils/responseParser');
const { RAGPipeline } = require('../rag/ragPipeline');

/**
 * AGENT 2: Skill Gap Analyzer
 *
 * Compares resume skills vs job description requirements.
 * Returns: matchedSkills, missingSkills, matchPercentage
 */
const runSkillGapAgent = async (ragPipeline) => {
  console.log('\n🤖 Agent 2: Skill Gap Analyzer running...');

  const resumeContext = await ragPipeline.retrieve(
    'technical skills programming languages frameworks tools technologies databases',
    6,
    'resume'
  );
  const jdContext = await ragPipeline.retrieve(
    'required skills must have technologies programming languages frameworks tools',
    6,
    'jd'
  );

  const systemPrompt = `You are a Technical Skill Gap Analyzer specializing in software engineering, data science, and AI roles. You excel at extracting and comparing technical and soft skills from resumes and job descriptions.`;

  const userPrompt = `Extract ALL skills from both documents and compare them.

RESUME:
${RAGPipeline.truncate(resumeContext, 2000)}

JOB DESCRIPTION:
${RAGPipeline.truncate(jdContext, 2000)}

Instructions:
- Extract all technical skills (languages, frameworks, tools, cloud, databases)
- Extract all soft skills (communication, leadership, teamwork)
- Identify which skills the candidate has vs what the job needs
- Calculate match percentage based on required skills covered

Return ONLY a JSON object:
{
  "matchedSkills": ["React", "Node.js", "MongoDB"],
  "missingSkills": ["Docker", "Kubernetes", "AWS"],
  "matchPercentage": <0-100>,
  "allRequiredSkills": ["all skills from JD"]
}`;

  const raw = await callGroq(systemPrompt, userPrompt, { temperature: 0.1 });
  const parsed = parseJSON(raw, {});
  return normalizeSkillGap(parsed);
};

module.exports = { runSkillGapAgent };

File: ai-service/agents/skill_gap_agent.py
"""
Agent 2: Skill Gap Analyzer Agent
====================================
Compares resume skills vs job description requirements.

Output:
    {
        "matchedSkills": [...],
        "missingSkills": [...],
        "matchPercentage": 75,
        "allRequiredSkills": [...]
    }
"""

from rag.rag_pipeline import RAGPipeline
from services.groq_service import call_groq
from utils.response_parser import parse_json, normalize_skill_gap
from prompts import skill_gap_prompt


async def run_skill_gap_agent(rag: RAGPipeline) -> dict:
    """
    AGENT 2: Skill Gap Analyzer

    Retrieves technical skills sections from both resume and JD,
    then compares them to identify matched and missing skills.

    Args:
        rag: Initialized RAGPipeline instance

    Returns:
        Normalized skill gap dict with matched, missing, and match percentage
    """
    print("\n🤖 Agent 2: Skill Gap Analyzer running...")

    # ─── RAG Retrieval ────────────────────────────────────────
    resume_context = await rag.retrieve(
        "technical skills programming languages frameworks libraries tools databases cloud",
        top_k=6,
        source="resume",
    )
    jd_context = await rag.retrieve(
        "required skills must have technical qualifications preferred technologies stack",
        top_k=6,
        source="jd",
    )

    # ─── Prompt Construction ──────────────────────────────────
    system_prompt = skill_gap_prompt.get_system_prompt()
    user_prompt = skill_gap_prompt.get_user_prompt(
        resume_context=RAGPipeline.truncate(resume_context, 2000),
        jd_context=RAGPipeline.truncate(jd_context, 2000),
    )

    # ─── LLM Call ─────────────────────────────────────────────
    raw = await call_groq(system_prompt, user_prompt, temperature=0.1)
    parsed = parse_json(raw, fallback={})
    result = normalize_skill_gap(parsed)

    print(f"   ✅ Match: {result['matchPercentage']}% | Missing: {len(result['missingSkills'])} skills")
    return result
